### PR TITLE
Allow path component in base URL

### PR DESCRIPTION
**Pull request by michaldepta**
This change allows using Authress through a proxy with a path component in its URL, e.g. https://example.com/authressProxyPath/

As described in https://tools.ietf.org/html/rfc3986#section-5.2.3 merging a base URI with a relative URI containing leading slash removes the base URI's path component.

### CODE REVIEW

**Comment by wparad** [comment]
Well this is certainly an interesting "bug".  The solution you have here seems to be consistent with SO articles as well, but I'm a bit concerned that it isn't sustainable.

For instance, someone may come along and add the slashes back, or generate new routes with problems.

Instead, (and full disclosure, Im not sure this actually doable, I haven't tried it out yet), can we create a message handler in the httpclientProvider and add a fix to always strip the leading slash.

That way we can't have a regression on this problem.  Additionally, since Authress won't ever use/expect explicit top level paths to overwrite the basepath, it seems like the right thing to do.

**Comment by michaldepta** [comment]
Good point. I don't think we can use message handlers for this purpose as they operate on already merged URIs.
One thing that can be introduced is some kind of URI factory method in `AuthressClient` and use it with `HttpClient`, e.g.:
```
        private static Uri CreateUri(string path) =>
            new Uri(path.StartsWith("/") ? path.Substring(1) : path, UriKind.Relative);
```
This won't ensure all new API calls are following the same convention, though.

**Comment by wparad** [comment]
> Good point. I don't think we can use message handlers for this purpose as they operate on already merged URIs.
> One thing that can be introduced is some kind of URI factory method in `AuthressClient` and use it with `HttpClient`, e.g.:
> 
> ```
>         private static Uri CreateUri(string path) =>
>             new Uri(path.StartsWith("/") ? path.Substring(1) : path, UriKind.Relative);
> ```
> 
> This won't ensure all new API calls are following the same convention, though.

I'm not sure where you would use the method without having to reference it in every api call.  If there is a way to make a "one-line" change somewhere, then okay!  (But I'm doubtful, prove me wrong)

While what you say about the message handler is true, we can rewrite the requestUri in a messagehandler.

On line 63 we can add a new handler
```cs
outermostHandler = new RewriteBaseUrlHandler(outermostHandler, settings.ApiBasePath);
```
And then add a new handler
```cs
internal class RewriteBaseUrlHandler : DelegatingHandler
    {
        private string baseUrl;
        public AddAuthorizationHeaderHandler(HttpMessageHandler innerHandler, string baseUrl) : base(innerHandler)
        {
            this.baseUrl = baseUrl;
        }

        protected override async Task<HttpResponseMessage> SendAsync(HttpRequestMessage request, CancellationToken cancellationToken)
        {
            request.RequestUri = new Uri(request.RequestUri.ToString().Replace(request.RequestUri.Host, baseUrl);
            return await base.SendAsync(request, cancellationToken);
        }
    }
```

**Comment by michaldepta** [comment]
Sure, this will work, although I'm not convinced if this is a proper usage of message handlers. 
Should I file another PR with such change?

**Comment by wparad** [comment]
Feel free to just update this one (that way we have all the tracking for the change in one place.)  Additionally with the message handler change, it should be much easier when it comes to adding the unit tests to the PR.

one caveat:
> request.RequestUri = new Uri(request.RequestUri.ToString().Replace(request.RequestUri.Host, baseUrl);

You'll need to figure out what this line should actually be, since it doesn't look like scheme + host is part of any of Uri object, and it also must end with a `/`.

> although I'm not convinced if this is a proper usage of message handlers.
Me either, but I'm not convinced the design team for message handlers knows what they are doing because they still haven't provided an interface for HttpClient which makes it easy to test.

Ultimately, I want to avoid a convention based approach since they are brittle, and I don't have a good suggestion to do that without using the message handler.

Additionally, there are some API methods that work on absolute urls, using the hypermedia linking. I assume you'll want to "rewrite" those too. In which case we **must** use the message handler there isn't an alternative.

**Comment by michaldepta** [comment]
@wparad please have a look now. With the URL rewriting handler I might as well revert the removal of leading slashes. What do you think?

**Comment by wparad** [comment]
> With the URL rewriting handler I might as well revert the removal of leading slashes. What do you think?

Yeah, we won't need that change anymore.

File: src/Authress.SDK/Client/HttpClientProvider.cs [hunk 1]
```diff
@@ -107,4 +108,30 @@ protected override async Task<HttpResponseMessage> SendAsync(HttpRequestMessage
...
         }
     }
+
+    internal class RewriteBaseUrlHandler : DelegatingHandler
+    {
+        private readonly Uri baseUrl;
+
+        public RewriteBaseUrlHandler(HttpMessageHandler innerHandler, string baseUrl) : base(innerHandler)
+        {
+            this.baseUrl = new Uri(baseUrl.EndsWith("/") ? baseUrl : baseUrl + "/");
```
Comment (wparad) [inline src/Authress.SDK/Client/HttpClientProvider.cs:118]: I just found `TrimEnd` seems useful here. (`TrimStart` could be used an line 135 as well)

```suggestion
            this.baseUrl = new Uri(baseUrl.TrimEnd('/'));
```

**Comment by wparad** [comment]
> With the URL rewriting handler I might as well revert the removal of leading slashes. What do you think?

Agree, We don't need the change anymore here. Though, now it doesn't have any meaningful impact, so up to you what you want to do.

Also this is good to go by me, so depending on what you want to do with ^^, I can merge this.

**Review by wparad — APPROVED** [review]

**Comment by wparad** [comment]
https://www.nuget.org/packages/Authress.SDK/1.0.42

**Pull request merged**